File: app/src/main/java/com/hover/stax/schedules/ScheduleDetailViewModel.java
package com.hover.stax.schedules;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;

import com.hover.sdk.actions.HoverAction;
import com.hover.stax.contacts.StaxContact;
import com.hover.stax.database.DatabaseRepo;

import java.util.List;

import static org.koin.java.KoinJavaComponent.get;

public class ScheduleDetailViewModel extends AndroidViewModel {

    private DatabaseRepo repo = get(DatabaseRepo.class);
    private MutableLiveData<Schedule> schedule;
    private LiveData<HoverAction> action;
    private LiveData<List<StaxContact>> contacts;

    public ScheduleDetailViewModel(@NonNull Application application) {
        super(application);
        schedule = new MutableLiveData<>();
        action = Transformations.switchMap(schedule, this::loadAction);
        contacts = Transformations.switchMap(schedule, this::loadContacts);
    }

    public void setSchedule(int id) {
        new Thread(() -> schedule.postValue(repo.getSchedule(id))).start();
    }

    public LiveData<Schedule> getSchedule() {
        if (schedule == null) {
            return new MutableLiveData<>();
        }
        return schedule;
    }

    private LiveData<HoverAction> loadAction(Schedule s) {
        if (s != null) {
            return repo.getLiveAction(s.action_id);
        }
        return new MutableLiveData<>();
    }

    public LiveData<HoverAction> getAction() {
        if (action == null) {
            action = new MutableLiveData<>();
        }
        return action;
    }

    private LiveData<List<StaxContact>> loadContacts(Schedule s) {
        if (s != null) {
            return repo.getLiveContacts(s.recipient_ids.split(","));
        }
        return null;
    }

    public LiveData<List<StaxContact>> getContacts() {
        if (contacts == null) {
            contacts = new MutableLiveData<>();
        }
        return contacts;
    }

    void deleteSchedule() {
        repo.delete(schedule.getValue());
    }
}

File: app/src/main/java/com/hover/stax/schedules/ScheduleDetailViewModel.kt
/*
 * Copyright 2022 Stax
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.hover.stax.schedules

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.switchMap
import androidx.lifecycle.viewModelScope
import com.hover.sdk.actions.HoverAction
import com.hover.stax.data.actions.ActionRepo
import com.hover.stax.data.schedule.ScheduleRepo
import com.hover.stax.database.models.Schedule
import com.hover.stax.database.models.StaxContact
import com.hover.stax.database.repo.ContactRepo
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class ScheduleDetailViewModel @Inject constructor(
    val repo: ScheduleRepo,
    val actionRepo: ActionRepo,
    val contactRepo: ContactRepo
) : ViewModel() {

    val schedule = MutableLiveData<Schedule>()
    var action: LiveData<HoverAction> = MutableLiveData()
    var contacts: LiveData<List<StaxContact>> = MutableLiveData()

    init {
        action = schedule.switchMap(this::loadAction)
        contacts = schedule.switchMap(this::loadContacts)
    }

    fun setSchedule(id: Int) = viewModelScope.launch { schedule.postValue(repo.getSchedule(id)) }

    private fun loadAction(s: Schedule?): LiveData<HoverAction> =
        if (s != null) actionRepo.getLiveAction(s.action_id) else MutableLiveData()

    private fun loadContacts(s: Schedule?): LiveData<List<StaxContact>> = if (s != null)
        contactRepo.getLiveContacts(s.recipient_ids.split(",").toTypedArray())
    else
        MutableLiveData()

    fun deleteSchedule() = repo.delete(schedule.value)
}